include/dynd/string.hpp

```diff
@@ -8,11 +8,52 @@
 #include <dynd/callable.hpp>
 
 namespace dynd {
-namespace nd {
 
-  extern DYND_API struct DYND_API string_concatenation : declfunc<string_concatenation> {
-    static callable make();
-  } string_concatenation;
+  /*
+    Concatenates `nop` strings in the `s` array, storing the result in
+    `d`
+  */
+  void string_concat(size_t nop, dynd::string *d, const dynd::string *const *s);
 
-} // namespace dynd::nd
+  /*
+    Returns the number of times needle appears in haystack.
+  */
+  intptr_t string_count(const dynd::string *const haystack,
+                        const dynd::string *const needle);
+
+  /*
+    Returns byte index of the first occurrence of needle in haystack.
+    Returns -1 if not found.
+  */
+  intptr_t string_find(const dynd::string *const haystack,
+                       const dynd::string *const needle);
+
+  /*
+    In string `src`, replace all non-overlapping appearances of
+    `old_str` with `new_str`, storing the result in `dst`.
+  */
+  void string_replace(dynd::string *const dst,
+                      const dynd::string *const src,
+                      const dynd::string *const old_str,
+                      const dynd::string *const new_str);
+
```

Comment:
My plan was to create APIs that take `dynd::string` or `std::string` that call lower-level APIs that take `char *`/`size_t` pairs (the C-like interface).

The detail I forgot in that plan is the memory management.  The C APIs will need to allocate and return the result buffer through a `char **` (since the caller won't know what size to allocate).  But then the C++ API will need to create a `dynd::string` out of it, without copying.  We could add an internal API to assign ownership of a pointer to a `dynd::string`, but such a thing doesn't exist in `std::string`, and it probably breaks @mwiebe's plans to add short-string optimization.

So, I think I come down on *just* supporting `dynd::string` and `std::string`, through a template, and not C-like buffers.  But maybe others have better ideas or strong opinions about what we *must* support here.